nixos/test-driver: Run commands with pipefail set

talyz opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Bash's standard behavior of not propagating non-zero exit codes through a pipeline is unexpected and almost universally
unwanted. Default to setting `pipefail` for the command being run; it can still be turned off by prefixing the pipeline with
`set +o pipefail` if needed.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Tested by constructing a test with a failing command early in the pipeline
- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](./CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

talyz committed:
nixos/test-driver: Run commands with error handling

Bash's standard behavior of not propagating non-zero exit codes
through a pipeline is unexpected and almost universally
unwanted. Default to setting `pipefail` for the command being run;
it can still be turned off by prefixing the pipeline with
`set +o pipefail` if needed.

Also, set `errexit` and `nonunset` options to make the first command
of consecutive commands separated by `;` fail, and disallow
dereferencing unset variables respectively.

tfc commented:
just out of curiosity, what kind of situation did you run into?

r-rmcgibbo commented:
Result of `nixpkgs-review pr 125683` at 548708c5 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.trivial</li>
  </ul>
</details>

talyz commented:
@tfc  A concrete example where running without this could create a false positive is in the test introduced in #125404, where if the `curl` command fails, the whole pipeline still succeeds because `jq` returns zero when given empty input. This should also apply to https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/discourse.nix and https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/keycloak.nix for the same reason.

tfc commented:
makes sense. can you please schedule a selection of tests that do some heavy piping between processes to see if we don't break anything? if a test breaks after such change then the test itself might be broken but we should handle that.

talyz committed:
nixosTests.nginx*: nginxUnstable -> nginxMainline

Stop using the old `nginxUnstable` alias, which is invalid in tests
since 3edde6562e19698da69a499881e0a2e4f5a497a2.

talyz commented:
Sure! I had to unbreak the `nginx` test for it to even evaluate.

@GrahamcOfBorg test cassandra_3_11 xandikos avahi  nginx docker-tools discourse keycloak networking prometheus-exporters ceph-multi-node ceph-single-node

talyz committed:
nixosTests.*: Don't use the `-q` flag with grep when used with curl

The `-q` flag makes grep close the pipe early, which curl doesn't
handle gracefully, but exits with an error like "(23) Failed writing
body".

roberth commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -441,7 +441,7 @@ def require_unit_state(self, unit: str, require_state: str = "active") -> None:
         self.connect()
 
-        out_command = "( {} ); echo '|!=EOF' $?\n".format(command)
+        out_command = "( set -o pipefail; {} ); echo '|!=EOF' $?\n".format(command)
```
```suggestion
        out_command = "( set -euo pipefail; {} ); echo '|!=EOF' $?\n".format(command)
```

Besides `false | echo ok`, I would expect `false; echo ok` and `echo $undefined` to fail.
Optional variables (unlikely in this use case) can still be used via `${undefined:-}` if necessary.

talyz replied:
Yeah, I agree. I've pushed the change and updated the manual.

talyz commented:
I've run lots of tests now and the problem I encountered was that when `grep -q` is used with `curl`, it closes the pipe early, which `curl` doesn't handle gracefully, but exits with an error, failing the pipeline. To fix this, I've simply removed the `-q` flag from `grep` where it's used with `curl`.

roberth commented:
The change it self lgtm, but I'm concerned that this may break tests in surprising ways for other who do not know about this change.
Was solving `grep -q` sufficient? How many tests can we expect to be broken still?

talyz commented:
Yes, solving the `grep -q` issue was enough for all the test I ran, and I ran quite a few, focusing on those who make heavy use of piping. I think `curl` is an outlier here, where most programs don't error when the pipe is closed early. Considering that all or almost all `curl` commands I could find have the `-f` flag specified, the test authors seem to have assumed that `pipefail` was already in effect, as I did when I wrote some of them.

roberth commented:
The change looked good, so I've kicked off https://hydra.nixos.org/eval/1676118 before merging, so we can find any new failures on hydra. It's currently evaluating with the change.

roberth commented:
I've started evaluation after the change, twice, but the evaluations disappeared before completion. I don't know why.
We'll have to check for newly failed tests when the next evaluation does complete.

talyz commented:
Hm, yeah, the error log at https://hydra.nixos.org/jobset/nixos/trunk-combined#tabs-errors contains lots of
```
warning: error: SQLite database '/nix/var/nix/db/db.sqlite' is busy
```
errors, but also ends with the rather cryptic
```
error: [json.exception.type_error.302] type must be string, but is null
```
so, maybe one of those is the culprit?

roberth commented:
Those errors are somewhat expected and happen during evaluation of nixpkgs. The tests evaluated fine and are listed [here](https://hydra.nixos.org/eval/1676118#tabs-now-fail) but many other changes were in that batch. I've checked most of them and they seem to fail for reasons unrelated to your pr.

talyz commented:
I see. The database warning doesn't appear in the successful evaluations I checked, so it seemed out of place, but I don't know much about Hydra :shrug:
